main/src/hadd.cxx

```diff
@@ -348,62 +381,127 @@ int main( int argc, char **argv )
+         if (argv[i] && argv[i][0]=='@') {
+            std::ifstream indirect_file(argv[i]+1);
+            if( ! indirect_file.is_open() ) {
+               std::cerr<< "hadd could not open indirect file " << (argv[i]+1) << std::endl;
+               return kFALSE;
             }
-         }
-      } else if( ! merger.AddFile(argv[i]) ) {
-         if ( skip_errors ) {
-            std::cerr << "hadd skipping file with error: " << argv[i] << std::endl;
-         } else {
-            std::cerr << "hadd exiting due to error in " << argv[i] << std::endl;
-            return 1;
+            while( indirect_file ){
+               std::string line;
+               if( std::getline(indirect_file, line) && line.length() &&  !merger.AddFile(line.c_str()) ) {
+                  return kFALSE;
+               }
+            }
+         } else if( ! merger.AddFile(argv[i]) ) {
+            if ( skip_errors ) {
+               std::cerr << "hadd skipping file with error: " << argv[i] << std::endl;
+            } else {
+               std::cerr << "hadd exiting due to error in " << argv[i] << std::endl;
+            }
+            return kFALSE;
          }
       }
-   }
-   if (reoptimize) {
-      merger.SetFastMethod(kFALSE);
-   } else {
-      if (!keepCompressionAsIs && merger.HasCompressionChange()) {
-         // Don't warn if the user any request re-optimization.
-         std::cout <<"hadd Sources and Target have different compression levels"<<std::endl;
-         std::cout <<"hadd merging will be slower"<<std::endl;
+      return mergeFiles(merger);
+   };
+
+   auto parallelMerge = [&](int start){
+      TFileMerger mergerP(kFALSE,kFALSE);
+      mergerP.SetMsgPrefix("hadd");
+      mergerP.SetPrintLevel(verbosity - 1);
+      if (maxopenedfiles > 0) {
+          mergerP.SetMaxOpenedFiles(maxopenedfiles/nProcesses);
       }
-   }
-   merger.SetNotrees(noTrees);
-   merger.SetMergeOptions(cacheSize);
+      if (!mergerP.OutputFile(partialFiles[(start-ffirst)/step].c_str(), newcomp)) {
+         std::cerr << "hadd error opening target partial file"<<std::endl;
+         exit(1);
+      }
+      return sequentialMerge(mergerP, start, step);
+   };
+
+   auto reductionFunc = [&](){
+      for(auto pf:partialFiles){
+         fileMerger.AddFile(pf.c_str());
+      }
+      return mergeFiles(fileMerger);
+   };
+
    Bool_t status;
-   if (append) status = merger.PartialMerge(TFileMerger::kIncremental | TFileMerger::kAll);
-   else status = merger.Merge();
+
+   if(multiproc) {
+      ROOT::TProcessExecutor p(nProcesses);
+      auto res = p.Map(parallelMerge, ROOT::TSeqI(ffirst, argc, step));
+      status = std::accumulate(res.begin(), res.end(), 0U) == partialFiles.size();
+      if(status){
+         status = reductionFunc();
+      } else{
+         std::cout<<"hadd failed at the parallel stage"<<std::endl;
+      }
+      for(auto pf:partialFiles){
+         gSystem->Unlink(pf.c_str());
```

Comment:
It could be useful, yes. 
How? Adding another option? This will only be used in the multiprocess case, so maybe a combination of the two? Something like -jdbg?